gpustack/policies/candidate_selectors/ascend_mindie_resource_fit_selector.py

```diff
@@ -0,0 +1,972 @@
+                )
+                return candidates
+            if (
+                self._model_params.num_attention_heads
+                % self._serving_params.data_parallel_size
+                != 0
+            ):
+                self._diagnostic_messages.append(
+                    f"Model's attention heads ({self._model_params.num_attention_heads}) must be divisible by "
+                    f"the data parallel size ({self._serving_params.data_parallel_size})."
+                )
+                return candidates
+
+        """
+        Get available workers.
+        """
+
+        ram_request, vram_request = estimated_usage[0], estimated_usage[1]
+
+        # Get available worker devices: {Worker: {Device Index: Device}},
+        # all devices are in sorted.
+        available_worker_devices_idx: Dict[Worker, Dict[int, GPUDeviceInfo]] = {}
+        # Record the remaining count needed to be selected,
+        # which is used to determine if enough devices are selected.
+        selected_devices_cnt_remain = self._serving_params.world_size
+
+        for worker in workers:
+            # Skip if the worker does not have devices.
+            if not worker.status.gpu_devices:
+                continue
+
+            # Choice devices of the worker based on the selected worker names and devices.
+            if self._selected_worker_name_devices_idx:
+                # Break if selected devices are enough,
+                # if the value is negative, it means not selected.
+                if selected_devices_cnt_remain == 0:
+                    break
+
+                # Skip if the worker is not in the selected worker names.
+                if worker.name not in self._selected_worker_name_devices_idx:
+                    continue
+
+                _ = await self.__get_worker_alloc(worker)
+
+                # Get selected devices of the worker: {Device Index: Device}.
+                selected_devices_idx: Dict[int, GPUDeviceInfo] = {
+                    device.index: device
+                    for device in self.__worker_sorted_devices_idx[worker.id]
+                    if device.index
+                    in self._selected_worker_name_devices_idx[worker.name]
+                }
+
+                # Skip if the worker does not have enough devices.
+                if len(selected_devices_idx) < len(
+                    self._selected_worker_name_devices_idx[worker.name]
+                ):
+                    continue
+
+                # Stats
+                selected_devices_cnt_remain -= len(selected_devices_idx)
+
+            # Otherwise, use all devices of the worker.
+            else:
+                worker_alloc = await self.__get_worker_alloc(worker)
+
+                # Skip if the worker does not have enough RAM.
+                ram_allocate = worker_alloc.ram
+                if ram_request > ram_allocate:
+                    continue
```

Comment:
Previously we ignore RAM. Now how do we tell users the cause when the scheduling fails due to insufficient RAM?

Reply:
there is an outside processing in `candidates = await self._construct_candidates(estimated_usage, workers)`, it will append a message about the resource requirements if there are no messages.